### PR TITLE
New implementation for currency on native app

### LINKED ISSUE #1772: Single Currency/Language for mobile app : same as web
As we have single currency/language mechanism for web, we need to develop same for mobile.
@sagararyal pls comment, design?

### LINKED ISSUE #1896: Page to Select Currency
Screen Shot 2020-02-04 at 7 18 47 PM

Flags are available at: CDN URL : /media/images/flags/svg/{country}.svg
https://cdn-app.plant-for-the-planet.org/media/images/flags/svg/DE.svg

Zeplin
https://zpl.io/bzXXjy7

Goal : This should replace dropdown which is inefficient and takes quite some time to load in the pages.

### PR DESCRIPTION
**Pull request by mahmed0715**
Fix #1896
Fix #1772
Implemented currency for native app. @sagararyal  as expected.
menu item on drawer left,
show featured/all currencies
used this as global currency
reflect this change to other places like currency selector in donation process/all plantproject pricess in donation process

if logged in used profile currency everywhere
onupdate the global currency if logged in update profile in db..

Issues:

- [x] replace constant 'EUR' with context.currency? https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1887#discussion_r381227903
- [x] use some default icon for `EUR` as this will be the chosen currency for most of our current users. Maybe @jmiridis can add the European flag to our assets? Maybe there is a nice € symbol in our icon font? Or we just use an emoji like 💶? But an empty space looks like something is missing.
- [x] clarify number formatting for different currencies and chosen language at https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1887#issuecomment-597142332
- [x] clarify border and borderless search field at https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1887#issuecomment-597146474
- [x] upload XC flag to CDNs https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1887#issuecomment-597662796

### CODE REVIEW

**Comment by mahmed0715** [comment]
Screenshot_1581437830
Screenshot_1581437825

@sagararyal pls see

**Comment by sagararyal** [comment]
@mahmed0715  the flags are not as they should be. the fonts and width also doesn't look as in design

**Comment by mahmed0715** [comment]
Screenshot_20200213-213125
i have done so far, I hope this is the final design. @sagararyal i had done that, please check once.

**Comment by mahmed0715** [comment]
"Design is not the same. It has padding around it in the app but not in zeplin design.
Check icon on the right is large"
@sagararyal need to check

**Comment by mahmed0715** [comment]
@ankitecd pls check https://github.com/Plant-for-the-Planet-org/treecounter-app/issues/1896

**Comment by ankitecd** [comment]
> @ankitecd pls check #1896

The issue mentions Zeplin please have a look at that screen.

**Comment by mahmed0715** [comment]
@ankitecd please remove the padding from the container and contact @sagararyal please..

**Comment by ankitecd** [comment]
@sagararyal I think the modal should have a back button instead of cross(x) on the left. The two cross (x) don't look good.

**Comment by norbertschuler** [comment]
- [ ] The API call https://test.trilliontreecampaign.org/api/v1.1/de/profileCurrency is sometimes taking quite long. Maybe the user show be shown an activity indicator to know he has to wait until the new currency is set and finally being selected in the list. We have deleted the old loading indicator for API calls by a PR of @haideralishah and not yet brought back something else for it, compare https://github.com/Plant-for-the-Planet-org/treecounter-app/issues/1837

**Comment by norbertschuler** [comment]
- [ ] Several times I clicked the "x" behind the search field trying to close the currency chooser. Maybe it would be better to only show the "x" if a filter is applied to the list or keep the surrounding box for the search field visible to avoid irritating users.

**Comment by norbertschuler** [comment]
- [ ] I find it irritating not seeing the chosen currency directly in the user menu. So you either have to open the currency chooser or open a project page to check what currency you have chosen. I'd rather like to see the chosen currency (e.g. the currency code like "EUR" or the country flag or both) directly within the menu item like it's done on the web client. I found this design about it, but don't know if it's still relevant (no, I don't think we need a language change functionality in the native apps): https://app.zeplin.io/project/5cad97d9b33d0b5931d90a9b/screen/5d36e59f2e8f1d41163248a7
- [ ] Also I think a green colored menu item icon does not fit to the other menu items and a pure black&white icon should be chosen if not showing a country flag of the chosen currency. 
- [ ] I don't think "EUR" has the country flag I currently see in the app, but rather https://en.wikipedia.org/wiki/Flag_of_Europe

File: app/actions/fetchConfig.js [hunk 1]
```diff
@@ -10,8 +10,10 @@ import { find } from 'lodash';
 import { setCurrencyAction } from './globalCurrency';
 // import { setCdnMedia } from '../reducers/configReducer';
 let cdnMedia = {};
+let currency = '';
 export function fetchLocation() {
   return dispatch => {
+    console.log('got preferred currency', getItemSync('preferredCurrency'));
```
Comment (norbertschuler) [inline app/actions/fetchConfig.js]: Please either delete your console.log again or use debug() as I replaced all console.log with debug lately.

File: app/components/Common/NumberFormat.native.js [hunk 2]
```diff
@@ -9,15 +9,20 @@ const NumberFormat = ({
...
+  currencies,
+  globalCurrency,
+  force
 }) => {
-  if (!userProfile && currency) {
+  if (!userProfile) {
     userProfile = {
-      currency: currency
+      // we use redux currency or use EUR if not found for now, dont use localstorage currency for now
+      currency: globalCurrency.currency || 'EUR'
```
Comment (norbertschuler) [inline app/components/Common/NumberFormat.native.js]: I'd rather use `context.currency` (need to import it) as a default value here.

**Comment by norbertschuler** [comment]
- [ ] I don't think "EUR" has the country flag https://bucketeer-894cef84-0684-47b5-a5e7-917b8655836a.s3.eu-west-1.amazonaws.com/test/media/images/flags/png/256/AS.png which I currently see in the app, but rather https://en.wikipedia.org/wiki/Flag_of_Europe (EU is not a country, so maybe you can't find a flag for it at all!)

File: app/components/Currency/GlobalCurrencySelector.native.js [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    if (!nextProps.currencies.currencies) {
+      this.props.fetchCurrencies();
+    }
+  }
+  async componentWillMount() {
+    // this.setState({ preferredCurrency: getPreferredCurrency() });
+    this.props.userProfile &&
+      this.rops.userProfile.currency &&
```
Comment (norbertschuler) [inline app/components/Currency/GlobalCurrencySelector.native.js]: Found this warning in the browser console:
```
Possible Unhandled Promise Rejection (id: 0):
TypeError: Cannot read property 'userProfile' of undefined 
at GlobalCurrencySelector.componentWillMount
```

**Comment by sagararyal** [comment]
@norbertschuler 

https://www.trilliontreecampaign.org/public/v1.1/en/currencies

is cached with cloudflare every 30 min, it shouldn't need a loading indicator.

**Comment by norbertschuler** [comment]
> @norbertschuler
> 
> https://www.trilliontreecampaign.org/public/v1.1/en/currencies
> 
> is cached with cloudflare every 30 min, it shouldn't need a loading indicator.

I think I mentioned https://test.trilliontreecampaign.org/api/v1.1/de/profileCurrency which is a different API call.

**Comment by sagararyal** [comment]
- [ ] Changing currency does not change the currency localization format

Screen Shot 2020-03-02 at 12 16 07 PM

It should be 1.08$ instead of 1,08$


- [ ] Currency Selector in Donate Screen hasn't been changed

Screen Shot 2020-03-02 at 12 14 51 PM

- [ ] Currency change in API can be done async, on the background. After selection redirect the user to profile instead of currency page.

Screen Shot 2020-03-02 at 12 17 23 PM

- [ ] When currency is being searched, do not show Featured and All, just show the search results.
in this case ILS is not a featured currency.
Screen Shot 2020-03-02 at 12 19 29 PM

- [ ] Follow design for search bar [current implementation is ugly]
Screen Shot 2020-03-02 at 12 20 42 PM

- [ ] Use full-width layout instead of specific margin
Screen Shot 2020-03-02 at 12 21 49 PM


- [ ] Icon is missing
Screen Shot 2020-03-02 at 12 22 44 PM

**Comment by norbertschuler** [comment]
> * [ ]  Changing currency does not change the currency localization format

@sagararyal I won't change the number formatting of currencies based on the chosen currency. In Germany I would except the numbers still to be formatted as German numbers even if another currency symbol is appended.

**Comment by norbertschuler** [comment]
> * [x]  Follow design for search bar [current implementation is ugly]

I am confused: Should there be a border around the search field (and closing button) now or not? (I prefer a border, but that's just my personal feeling.)

File: app/components/Common/CurrencySelectorList.native.js [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                  marginLeft: 20
+                }}
+              >
+                <TextInput
+                  style={{ height: 40, width: '84%' }}
+                  onChangeText={text => {
+                    this.setSearch(text);
+                  }}
+                  value={this.state.search}
+                  placeholder={'Search'}
```
Comment (norbertschuler) [inline app/components/Common/CurrencySelectorList.native.js]: Here a translation resource must be added.

File: app/components/Currency/GlobalCurrencySelector.native.js [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+          <Text
+            style={{
+              color: '#4D5153',
+              fontFamily: 'OpenSans-SemiBold',
+              fontSize: 18,
+              marginTop: 30,
+              textAlign: 'center'
+            }}
+          >
+            Setting up your default currency
```
Comment (norbertschuler) [inline app/components/Currency/GlobalCurrencySelector.native.js]: Here a translation resource must be added too.

File: app/utils/currency.js [hunk 6]
```diff
@@ -49,3 +49,23 @@ export const sortProjectsByPrice = (plantProjects, asc, currencies) => {
...
+
+/**
+ * Accepts currency array to sort the curriencies list and returns the same
+ * @param {array} currenciesArray, currencies array
+ */
+export const currencySort = currenciesArray => {
+  if (currenciesArray.indexOf('LBP') > 0) {
+    currenciesArray.splice(currenciesArray.indexOf('LBP'), 1);
+    currenciesArray.unshift('LBP');
+  }
```
Comment (norbertschuler) [inline app/utils/currency.js]: ;-)

**Comment by harshvitra** [comment]
> > * [x]   Follow design for search bar [current implementation is ugly]
> 
> I am confused: Should there be a border around the search field (and closing button) now or not? (I prefer a border, but that's just my personal feeling.)

@ankitecd 
Add thin underline under the search bar when active

**Comment by harshvitra** [comment]
> @sagararyal I won't change the number formatting of currencies based on the chosen currency. In Germany I would except the numbers still to be formatted as German numbers even if another currency symbol is appended.

We will discuss later, now we can push without this one

**Comment by harshvitra** [comment]
> * [ ]  use some default icon for `EUR` as this will be the chosen currency for most of our current users. Maybe @jmiridis can add the European flag to our assets? Maybe there is a nice € symbol in our icon font? Or we just use an emoji like 💶? But an empty space looks like something is missing.
> *

We can add icons in the backend ourselves 
https://github.com/Plant-for-the-Planet-org/treecounter-platform/tree/develop/web/flags

File: app/components/Common/NumberFormat.native.js (hunk 2 above)
Comment (norbertschuler) [reply app/components/Common/NumberFormat.native.js]: Did it myself now.

**Comment by norbertschuler** [comment]
> We can add icons in the backend ourselves
> https://github.com/Plant-for-the-Planet-org/treecounter-platform/tree/develop/web/flags

I added the EU flag and @sagararyal has distributed it to all CDNs. Unfortunately loading the PNG files of all countries produced quite some traffic. SVG would be better, having the flags locally maybe best.

**Comment by norbertschuler** [comment]
- [ ] The east caribbean dollar (XCD) has the same problem as EUR as no it's used for several countries and I don't know of any flag for it. Currently the app tries to load a file named `XC.png`, but we don't have it in our assets.

**Comment by norbertschuler** [comment]
@sagararyal can you help deciding or solving the last two open issues, so we can merge this PR?

**Comment by sagararyal** [comment]
todo:

- [ ] Fix for Header
replace cross with back button when search is initiated. [so the cross on right can be removed]

- [ ] Sort All Currencies alphabetically

**Comment by norbertschuler** [comment]
> * [ ]  Sort All Currencies alphabetically

@sagararyal Do you mean sorting by currency symbol or currency name? This would be sorting by currency symbol with German currency names:

Simulator Screen Shot - iPhone 11 - 2020-03-21 at 22 08 40

**Pull request merged**